lesson12/index.js

```diff
@@ -0,0 +1,78 @@
+const contents = document.getElementById('js-contents');
+const button = document.createElement('button');
+
+const displayLoadingImage = () => {
+    const loadingImage = document.createElement('img');
+    loadingImage.id = 'loadingImageId';
+    loadingImage.src = './loading-circle.gif';
+    contents.appendChild(loadingImage);
+}
+
+const removeLoadingImage = () => {
+    const loadingImage = document.getElementById('loadingImageId');
+    loadingImage.remove();
+}
+
+const getData = async (url) => {
+    try {
+        const response = await fetch(url);
+        return await response.json();
+    } catch (error) {
+        throw new Error(error);
+    }
+};
+
+const init = async () => {
+    displayLoadingImage();
+    removeButton();
+    try {
+        const { data } = await getData("https://myjson.dit.upm.es/api/bins/edwr");
+        if (data.length === 0) {
+            errorText(`エラーが発生しました: 詳細 data is empty`);
+            return;
+        }
+        createList(data);
+    } catch (e) {
+        errorText(`エラーが発生しました: 詳細 ${e}`);
+    }
+    removeLoadingImage();
+}
+
+const errorText = (text) => {
+    const errorText = document.createElement('p');
+    contents.replaceWith(errorText);
+    errorText.textContent = text;
+}
+
+const createList = (lists) => {
+    const frag = new DocumentFragment();
+    const ul = document.createElement('ul');
+    for(const list of lists) {
+        const li = document.createElement('li');
+        const a = document.createElement('a');
+        a.href = list.a;
+        const img = document.createElement('img');
+        img.src = list.img;
+        img.alt = list.alt;
+        const text = document.createTextNode(list.text);
+
+        a.appendChild(img);
+        a.appendChild(text);
+        li.appendChild(a);
+        ul.appendChild(li);
+        frag.appendChild(ul);
+    }
+    contents.appendChild(ul);
+}
```

Comment:
[About this one.](https://github.com/kanakonagiri/js-lessons/pull/12#discussion_r757745371)

```js
        ul.appendChild(li);
        frag.appendChild(ul);
```
[Document.createDocumentFragment()](https://developer.mozilla.org/ja/docs/Web/API/Document/createDocumentFragment)
According to this.
Document.createDocumentFragment() makes sense if you put li in the Fragment you created in the loop beforehand; if you put ul in the Fragment, it doesn't make sense to use it.

IMO
Put the li in the fragment inside the loop process, and put the ul in the fragment created outside the loop process.
After that, if we put the ul inside the content, there will be no extra processing, won't there?
```suggestion
const createList = (lists) => {
    const frag = new DocumentFragment();
    const ul = document.createElement('ul');
    for(const list of lists) {
        const li = document.createElement('li');
        const a = document.createElement('a');
        a.href = list.a;
        const img = document.createElement('img');
        img.src = list.img;
        img.alt = list.alt;
        const text = document.createTextNode(list.text);

        a.appendChild(img);
        a.appendChild(text);
        li.appendChild(a);
        frag.appendChild(li);
    }
    contents.appendChild(ul).appendChild(frag);
}
```

Reply:
I didn't understand the meaning of using Document Fragment.
I received a review and was able to understand!
Thank you very much.
I fixed -> 2f5982d

Reply:
Thanks for the fix.　Approved!

In my previous review, I showed code that uses method chaining.
So there is currently a mixture of parts that use method chaining and parts that do not.

The next challenge is to use a method chain to
Here.
```js
a.appendChild(img);
a.appendChild(text);
 li.appendChild(a);
frag.appendChild(li);

```
Try to implement it using a method chain, like this code!
```js
contents.appendChild(ul).appendChild(frag);
```

Reply:
I put two elements in the a tag with appendChild, so I left it as it is, but is it correct?
-> 77e5277

Reply:
That's so nice and Excellent.
I'd use this one.
```js
insertAdjacentElement()
```
You can use it like this
```js
const createList = (lists) => {
  const frag = new DocumentFragment();
  const ul = document.createElement("ul");
  for (const list of lists) {
    const li = document.createElement("li");
    const a = document.createElement("a");
    a.href = list.a;
    a.textContent = list.text;
    const img = document.createElement("img");
    img.src = list.img;
    img.alt = list.alt;

    frag
      .appendChild(li)
      .appendChild(a)
      .insertAdjacentElement("afterbegin", img);
  }
  contents.appendChild(ul).appendChild(frag);
};
```
Delete here.
```js
 const text = document.createTextNode(list.text);
```
I put the text directly into `a`.
```js
 a.textContent = list.text;
```
Click here to see how it works.
[codesandbox](https://codesandbox.io/s/charming-tereshkova-e67vb?file=/index.js)

Reply:
Thank you so much!!!
I was able to rewrite it into very good code!!!
-> 308e997